pkg/controller/certificates/issuing/issuing_controller_test.go

```diff
@@ -174,6 +174,38 @@ func TestIssuingController(t *testing.T) {
 			expectedErr: false,
 		},
 
+		"if certificate is in Issuing state, one CertificateRequest, but has failed and does not match the certificate spec, do nothing": {
+			certificate: exampleBundle.Certificate,
+			builder: &testpkg.Builder{
+				CertManagerObjects: []runtime.Object{
+					issuingCert.DeepCopy(),
+					gen.CertificateRequestFrom(createCertificateRequestOrPanic(gen.CertificateFrom(issuingCert,
+						gen.SetCertificateDNSNames("foo.com"), // Mismatch since the cert has "example.com"
+					)), gen.AddCertificateRequestAnnotations(map[string]string{
+						cmapi.CertificateRequestRevisionAnnotationKey: "2", // Current Certificate revision=1
```

Comment:
I'm struggling to understand how there can be a CertificateRequest with the correct revision annotation but which does not match the Certificate. Does it go like this:

* Save broken Certificate.Spec
* CertificateRequest created but fails
* Fix Certificate.Spec 
* CertificateRequest already exists with the correct revision annotation, but broken spec


I think I probably don't understand how all these controllers interact.
I also can't help thinking that the CertificateRequest should be matched to a particular Certificate.ObjectMeta.Revision so that the controllers can quickly abandon CertificateRequests which are for an out-of-date Certificate.

Perhaps that's something for the future though.

Reply:
Totally agree, this test case is quite confusing 😞 Yes, it goes exactly as you say: after updating the certificate, the previous request might still be around.

To understand this test case, let's first study the "happy path" update case. Let's imagine the certificate is already at revision 1. When no failure happens on the request created in step (2), the request gets deleted by the issuing controller in step (4); since the previous request has been properly deleted, the certificate update in (5) works as expected:

| step | event              | cert             | request          |
| ---- | ------------------ | ---------------- | ---------------- |
| 0    | —                  | rev = 1, ready   | (none)           |
| 1    | cert updated       | rev = 1, issuing | (none)           |
| 2    | req created        | rev = 1, issuing | rev = 2, pending |
| 3    | req becomes ready  | rev = 1, issuing | rev = 2, ready   |
| 4    | cert becomes ready | rev = 2, ready   | (deleted)        |
| 5    | cert is updated    | rev = 1, issuing | (none)           |
| 6    | req created        | rev = 1, issuing | rev = 2, pending |

Now, imagine that the request fails created in (2). In this case, the request created in step (2) is left over and stays after step (4): the issuing controller picks up the old request and shouldn't do anything with it:

| step | event               | cert             | request                        |
| ---- | ------------------- | ---------------- | ------------------------------ |
| 0    | —                   | rev = 1, ready   | (none)                         |
| 1    | cert updated        | rev = 1, issuing | (none)                         |
| 2    | req created         | rev = 1, issuing | rev = 2, pending               |
| 3    | req becomes failed  | rev = 1, issuing | rev = 2, failure               |
| 4    | cert becomes failed | rev = 1, failure | rev = 2, failure               |
| 5    | cert updated        | rev = 1, issuing | rev = 2, failure (**old req**) |

That's why we end up with a test case that has `Issuing = True` and a failing request that doesn't match the certificate 😅

Reply:
> I also can't help thinking that the request should be matched to a particular `Certificate.ObjectMeta.Revision` so that the controllers can quickly abandon requests which are for an out-of-date certificate.

Good point! We could use a better way of knowing when the certificate and the request are out-of-sync.

Using the `metadata.revision` to select the correct request would be a good starting point, the only problem being that the `metadata.revision` changes every time the object changes (including statuses, which might often change). We would like to "select" on something that would be specific to the certificate's spec content rather than the whole object.

The Kubernetes codebase uses two mechanisms to do this: the `metadata.generation` and the hashing approach.

## The `metadata.generation` approach

All the objects containing the `spec` subresource also have a `metadata.generation` field; it is incremented any time the spec gets changed.

If we think that any kind of change in the certificat's spec should be triggering a new request issuance, then we could use this approach. For example,

```yaml
kind: Certificate
metadata:
  generation: 4 # Apiserver increments this when the cert's spec changes.
---
kind: CertificateRequest
metadata:
  labels:
    cert-generation: 4
```

- **Pros**: the "spec change" detection is already dealt with by Kubernetes. Since the generation is stored as a label, looking for matching requests is blazing fast (just using selectors while listing). Still works when the child object (here, the request) is defaulted/mutated in any way.
- **Cons**: if we want to "ignore" the update of some field in the certificate spec, we are out of luck. If the child gets updated, the parent cannot know that it has been changed since the hash only works in one way.

## The hashing approach

For example, the replica set controller in Kubernetes uses a hashing mechanism ([`ComputeHash`](https://github.com/kubernetes/kubernetes/blob/7e75a5ef/pkg/controller/controller_utils.go#L1130-L1145) which uses [fnv.New32a](https://golang.org/pkg/hash/fnv/#New32) from the stdlib) to know when the `spec.template` of a pod is out-of-sync with its replicaset's spec.

Pods that are owned by a replica set have a label that is used to remember what the hash of the pod template that led to this pod spec:

```yaml
kind: Pod
metadata:
  labels:
    pod-template-hash: 775855699b
```

- ✅ **Pros**: Allows you to ignore some fields in the parent's spec. Offers a workaround to the fact that `reflect.DeepEqual` can't work when the children objects are getting mutated or defaulted.
- ❌ **Cons**: if the child gets updated, the parent cannot know that it has been changed since the hash only works in one way.

## Conclusion

Currently, we use [RequestMatchesSpec](https://github.com/jetstack/cert-manager/blob/6fd14b02410e5a55cd5054d70ed783dca169f503/pkg/controller/certificates/util.go#L97-L164) for detecting when the certificate's spec and the request are out-of-sync. It looks like all the certificate fields are being "checked" (not sure), so we might be better off with the `metadata.generation` approach. But if we think some fields of the certificate's spec shouldn't be used to check for equality, then the hashing mechanism would make more sense.

Reply:
Great write up! Thanks. And the table of Cert, CR interactions makes it clear to me the situation we're fixing here.

And thanks for explaining the generation vs revision thing...I'd forgotten about that.